src/core/layertree/qgslayertreemodel.h

```diff
@@ -149,12 +150,28 @@ class CORE_EXPORT QgsLayerTreeModel : public QAbstractItemModel
     double legendFilterByScale() const { return mLegendFilterByScale; }
 
     //! Force only display of legend nodes which are valid for given map settings.
+    //! The selection of legend nodes can be refined by passing a boolean expression and/or a polygon (added in 2.12).
     //! Setting null pointer or invalid map settings will disable the functionality.
+    //! If the expression is not empty and valid, only legend nodes where the expression is evaluated to true for at least one feature are displayed.
+    //! If the polygon is not null, only legend nodes where symbols are inside the polygon are displayed. Coordinates of the polygon are assumed to be
+    //! expressed in the CRS of the map settings.
     //! Ownership of map settings pointer does not change, a copy is made.
+    //! Ownership of the expression does not change, it is a (possibly null) reference
+    //! Ownership of the polygon does not change, it is copied
     //! @note added in 2.6
-    void setLegendFilterByMap( const QgsMapSettings* settings );
+    void setLegendFilterByMap( const QgsMapSettings* settings, /*const*/ QgsExpression* expression = 0, const QgsGeometry* polygon = 0 );
```

Comment:
I think the expression should be copied too, to avoid crashers where someone will pass a temporary pointer (e.g. expression instance on stack). Also, why not just use QString with expression text?